browser/autocomplete/search_suggestions_query_check_utils.cc

```diff
@@ -0,0 +1,420 @@
+  //
+  // * If the path name itself gets too long, we have to careful
+  //   because of links from URL shortener (e.g., 'bit.ly/1h0ceQI').
+  // * If it contains URL search paramters in addition to the
+  //   path ('<domain>/path?param[=key]'), also stop.
+  //
+  // Note: ".search" without ".pathname" is quite aggressive,
+  // for instance, 'Fu?ball' (misspelling for "Fußball"), would
+  // already match ("http://Fu?ball" => host: "Fu", search: "ball")
+  if (url.path() != "/" && (url.path().length() >= 6 || url.has_query())) {
+    return true;
+  }
+
+  const auto domain_guessable = RE2::PartialMatch(query, "\\w+[.]\\w+\\/");
+  if (domain_guessable) {
+    return true;
+  }
+
+  // looks safe
+  return false;
+}
+
+}  // namespace
+
+namespace search_suggestions {
+
+bool IsSuspiciousQuery(const std::string& query) {
+  if (!HasValidQueryLength(query)) {
+    return true;
+  }
+
+  if (!HasValidWordCountInQuery(query)) {
+    return true;
+  }
+
+  if (HasLongNumberInQuery(query)) {
+    return true;
+  }
+
+  if (HasEmailInQuery(query)) {
+    return true;
+  }
+
+  if (LooksLikeHttpPass(query)) {
+    return true;
+  }
+
+  if (CheckHashProb(query)) {
+    return true;
+  }
+
+  return false;
+}
+
+bool IsSafeQueryUrl(const std::string& query) {
+  // fast path, which should handle most calls
+  const auto is_small_enough = [](const std::string& query) {
+    return query.length() <= 6;
+  };
+
+  if (is_small_enough(query)) {
+    return true;
+  }
+
+  if (query.length() > 500) {
+    // If the query is extremely large, abort immediately and do
+    // not attempt to normalize the query first, which can be a
+    // costly operation depending on the size of the text.
+    //
+    // This limit should be rather conservative, as the real
+    // check for huge texts operates on the normalized data.
+    // Here, we only want to make sure that the browser does not
+    // unnecessarily freeze if you paste a multi MB text file.
+    return false;
+  }
+
+  std::string updated_query = query;
+  RE2::GlobalReplace(&updated_query, "\\s+", " ");
+  if (is_small_enough(updated_query)) {
+    return true;
+  }
+
+  // Do not attempt to search for long texts. Maybe the user
+  // accidentally copied a sensitive email to the clipboard.
+  //
+  // However, if the limit is too low, we miss valid queries
+  // when someone copies error messages and searches for it.
+  //
+  // Note: In Feb 2018, around 99.9% of searches are less
+  // than 100 characters.
+  //
+  // (TODO: If the text uses a bigger alphabet, we should enforce
+  // a stricter limit. Extreme case is Chinese, where a limit
+  // of 20-30 character/words could make more sense.)
+  if (updated_query.length() > 100) {
+    return false;
+  }
+
+  // http[s] is a special case, but all other specific protocols
+  // are potentially dangerous, so immediately exit. Also it is
+  // quite unlikely that we would get good search results.
+  if (IsBlacklistedSchema(query)) {
```

Comment:
I believe we don't need this part because we whitelist domains already via Chromium's `IsQueryPotentiallyPrivate()`: https://source.chromium.org/chromium/chromium/src/+/main:components/omnibox/browser/search_provider.cc;l=801-804;drc=bbb50045f86f0e7c02c36b9ac7fddcb8faddfee9